src/bin/snapshots.rs

```diff
@@ -0,0 +1,122 @@
+extern crate crossbeam;
+extern crate gll;
+extern crate insta;
+extern crate proc_macro2;
+extern crate rayon;
+extern crate rust_grammar;
+extern crate walkdir;
+extern crate regex;
+extern crate lazy_static;
+
+use {
+    insta::assert_snapshot_matches,
+    rust_grammar::parse,
+    std::{fmt::Debug, fs, process::exit},
+    walkdir::WalkDir,
+    regex::Regex,
+    lazy_static::lazy_static,
+};
+
+fn to_debug_str(debug: &dyn Debug) -> String {
+    format!("{:#?}", debug)
+}
+
+macro_rules! snapshot {
+    ($production:ident, $src:expr) => {
+        match $src.parse::<proc_macro2::TokenStream>() {
+            Ok(tts) => parse::$production::parse_with(tts, |_, result| to_debug_str(&result)),
+            // FIXME(eddyb) provide more information in this error case.
+            Err(_) => to_debug_str(&Err::<(), _>(parse::ParseError::<()>::NoParse)),
+        }
+    };
+}
+
+macro_rules! dispatch {
+    ($src:ident, $s:expr; $($prod:ident)*) => {
+        match $s {
+            $(stringify!($prod) => snapshot!($prod, $src),)*
+            prod => panic!("Unexpected production {} tested", prod),
+        }
+    };
+}
+
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"\(\s+_,?\s+\)").unwrap();
+}
+
+fn test_snapshot(file: walkdir::DirEntry) {
+    let path = file.path();
+    let file_name = file.file_name().to_str().unwrap();
+    let src = fs::read_to_string(path).unwrap();
+    let production = &file_name[..file_name.find('.').unwrap_or(file_name.len())];
+    let forest = dispatch! { src, production;
```

Comment:
Seems rather unfortunate that you have to change the testing infra when you move or add new productions.

Reply:
Unfortunately without adding _another_ layer of proc-macros (which _is_ doable but seems silly) this is the best way I could think of to turn the strings into type dispatch. And they don't even all implement a trait, it's just a shape.

Note that the testing infra only has to change to _test_ new productions; we could remove all the productions from the list that aren't being tested. E.g. we'd add the production to the list and test it to "stabilize" its contents. (🤔 This actually makes a lot of sense)

Reply:
Fair enough; could you document what the consequence of adding or removing a production from the list is then with a comment? Maybe also update the README?

Reply:
You could generate this in `build.rs`, I think? If we have to, we can add an API to `Grammar` to list all the named rules.

Reply:
We'll need some way to access the list of rules from `build.rs`, so that'd be an API on `Grammar` to get the list. Once we do that we can output the `dispatch!` call to `OUT_DIR` and `include!` it from there.